microsoft/testsuites/mdatp/mdatp.py

```diff
@@ -0,0 +1,125 @@
+    description="""
+        Test to verify there are no pre-installed copies of mdatp.
+    """,
+)
+class MdatpSuite(TestSuite):
+    @TestCaseMetadata(
+        description="""
+           Check for mdatp endpoint/cloud install, dump config info.
+           Fails if mdatp is installed in the image.
+           Raises specific error messages depending on the type of info
+           foud
+        """,
+        priority=3,
+        requirement=simple_requirement(
+            unsupported_os=[Windows], supported_platform_type=[AZURE]
+        ),
+    )
+    def verify_mdatp_not_preinstalled(self, node: Node) -> None:
+        # collect some paths before we start the test
+        checker = "check-mdatp.sh"
+        local_path = pathlib.PurePath(__file__).parent.joinpath(checker)
+        working_path = node.get_working_path()
+        script_path = working_path.joinpath(checker)
+
+        # copy the bash script to the node
+        node.shell.copy(local_path=local_path, node_path=script_path)
+        result = node.execute(
+            cmd=f"chmod +x {str(script_path)}",
+            shell=True,
+            sudo=True,
+        )
+        # and run the script to check and dump defender info
+        result = node.execute(
+            cmd=f"{str(script_path)}",
+            shell=True,
+            sudo=True,
+            no_debug_log=True,
+        )
+        script_output = result.stdout.strip()
+        exit_code = result.exit_code
+        if exit_code is None:
+            raise SkippedException("exit code was None after running check-mdatp")
+        if exit_code == 0:
+            node.log.info("No mdatp onboarding info found, image is clean.")
+            return
+
+        node.log.info("Found mdatp installation!")
+        node.log.debug(f"Found config information: {script_output}")
+        found_onboarding_info = exit_code & EXIT_ONBOARD_INFO_FOUND
+        found_install_logs = exit_code & (
+            EXIT_MDATP_INSTALL_LOGS_FOUND | EXIT_MDATP_LOGS_FOUND
+        )
+        found_mdatp_installed = exit_code & EXIT_MDATP_AGENT_INSTALLED
+        found_mde_extension_install = exit_code & EXIT_MDE_INSTALLED
+
+        # Add some descriptive text to describe each specific problem.
+        error_header = ""
+        if found_onboarding_info:
+            error_header += "mdatp onboarding info is present in this image! "
+        if found_install_logs:
+            error_header += "mdatp install logs are present in this image! "
+        if found_mdatp_installed:
+            error_header += "mdatp installation was found on this image! "
+        if found_mde_extension_install:
+            error_header += "MDE extension installation was found in this image! "
+        if not any(
+            [
+                found_onboarding_info,
+                found_install_logs,
+                found_mdatp_installed,
+                found_mde_extension_install,
+            ]
+        ):
+            raise SkippedException(
```

Comment:
Should it mean passed?

Reply:
it needs at least one path to be Passed, or there is no way to calculate pass rate, evaluate the quality of the test case.

Reply:
I think we covered this already, there is a path to passing.

Reply:
From the message, "no recognized error code found." if no error, it should be passed, why skip it? If as you said, it won't happen, if it happens, that's something unexpected. It should be failed. It doesn't make sense to skip here.

Reply:
Does SKIP make an image fail in AzCertify or something? 
If we get an unexpected error code it implies we're running on a non-posix system, like maybe we've ended up on ReactOS or something. I think SerenityOS has a posix shell so that one will probably work, even though defender isn't supported. I don't feel good about giving a pass when we haven't found what we're looking for and the script probably hasn't even run correctly.

I changed this code path because of feedback from you, it was a hard fail if the return code was nonzero. You said this needs to be 100% reliable to use with AzCertify, so I modified the script such that we don't use bash-specifics, we look for specific things in the shell script, we don't rely on string parsing, and we allow for a skip case in the event we can't validate anything correctly.

I've taken time to make sure this shell script works with all unix shells including BSD, XNU/MacOS/OSX, and Linux.  If this returns anything but the 5 error codes in the script: we're not running in an OS which supports defender **as of 4/11/2024**. 

If we're assuming this test is going to be around longer than 10 years, I don't think we should give new unexpected OS's or unexpected shells the same passing grade as a known, validated OS (or non-compliant shell).

Having said that, you have made me rethink the error codes. It would probably be better to return something higher than 127 so we don't conflict with errno? I'm going to change that, it would be nice to get real exit codes for OS errors. Then we can change the message to "Unexpected exit code found: "

Reply:
Fixed 🙂

Reply:
By the way, is there a way to raise a PASS with a failure message? I find I often want to add a 'pass' note and it would be helpful for analyzing results after the test run. We can discuss offline.

Reply:
@squirrelsc I was wrong, after re-writing this skip message... it should be a fail! Thanks for chatting w me offline and helping me understand your thought process, concerns, and what actions were needed.

Reply:
My thought is, if there's suddenly a new posix-shell incompatible image in the marketplace... we need to take this into account. We also don't want to allow people to skip the check by fiddling with the shell defaults.